sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -525,18 +596,24 @@ public void populateDisplayData(DisplayData.Builder builder) {
 
     private ReadFn(
         DataSourceConfiguration dataSourceConfiguration,
+        DataSourceFactory dataSourceFactory,
         ValueProvider<String> query,
         PreparedStatementSetter<ParameterT> parameterSetter,
         RowMapper<OutputT> rowMapper) {
       this.dataSourceConfiguration = dataSourceConfiguration;
+      this.dataSourceFactory = dataSourceFactory;
       this.query = query;
       this.parameterSetter = parameterSetter;
       this.rowMapper = rowMapper;
     }
 
     @Setup
     public void setup() throws Exception {
-      dataSource = dataSourceConfiguration.buildDatasource();
+      if (dataSourceFactory != null) {
+        dataSource = dataSourceFactory.create();
```

Comment:
Don't we want to wrap the result of `dataSourceFactory` in a pool?

Reply:
In the case where the user provides its own `dataSourceFactory`, I guess he wants to have complete control and can "wrap" himself in a pool no ? The pooling is in the case of the user provides a `dataSource` or configuration to create a `dataSource` internally in the IO (using DBCP). Thoughts ?

Reply:
Can you think of a case where the current default pooling is not a good choice AND a user can provide a pooling configuration that will reliably give better behavior regardless of runner and of implementation details of JdbcIO?

Reply:
In other words: In my view, dataSourceFactory is rather for cases where the DataSource requires some trickery to even construct, e.g. requires setting some vendor-specific connection parameters to establish the connection. I don't know if such cases exist at all, and I'd be open to simply dropping this feature from this PR. But if we think it's necessary, then I think we still want to wrap that with pooling, and I think for a user to try to configure pooling themselves is always a bad idea.

Reply:
Maybe for backward compatibility, it would be better to still support `dataSourceFactory`, right ? In that case, I would "wrap" `dataSourceFactory` in the pool as you suggested previously. Agreed ?

Reply:
What do you mean by backwards compatibility? This *is* the PR that adds `dataSourceFactory`, so not adding it in this PR does not break compatibility.

Reply:
Oh sorry, you are right. I thought I added `dataSourceFactory` in another PR ;) In that case, I agree with you, I can be removed ;) I do the cleanup. Thanks !